### ISSUE TITLE
Slow querying on large documents.

### ISSUE THREAD

**Issue by kwando** [rethinkdb/rethinkdb#5023]
I'm trying out storing binary data in rethinkdb, while it works performance is less than stellar. In table with 218 docs and a combined filesize of ~610MB.

Avg size: 2.8MB
Std size: 3.1
Max size: 24.0MB

Sample doc
```ruby
{
  "id"=>"bad172fe-3340-4cbe-94be-572413f7e07a", 
  "name"=>"HHEvkhEsQLaNDRkui4fg_fullsize.jpg", 
  "size"=>9852159,
  "data"=><Binary data>
}
```
# 1 shard && 1 replica
```
r.db("merciless").table("uploads").order_by({"index" => r.desc("size")}).without("data").limit(10)
456.9ms
```

```
r.db("merciless").table("uploads").count
21.9ms
```

```
r.db("merciless").table("uploads").sum("size")
427.1ms
```

# 2 shards && 2replicas (same machine)
```
r.db("merciless").table("uploads").order_by({"index" => r.desc("size")}
).without("data").limit(10)
4578.8ms
```
```
r.db("merciless").table("uploads").count
2.6ms
```
```
r.db("merciless").table("uploads").sum("size")
640622041
3143.1ms
```

If I drop the data field ```table.update({data: r.literal()})``` all queries return in ~5ms.

```
$ rethinkdb --version
rethinkdb 2.1.5-2 (CLANG 7.0.0 (clang-700.0.
```
```
ruby 2.2.3p173 (2015-08-18 revision 51636) [x86_64-darwin14] with driver version 2.1.0.2
```

**Comment by danielmewes** [comment]
Thanks for reporting this @kwando .

RethinkDB is mostly optimized for documents of up to a couple 100 KB in size. That being said there's no technical reason as far as I can tell that would stop us from improving performance for larger documents like yours. Especially the slowdown when going to two shards (even though it's on the same machine) is unexpected.

We'll investigate this and see what we can do to optimize performance in the server.

**Comment by kwando** [comment]
Yeah, I know it is not the typical workload for RethinkDB.  I was just hoping save myself from some work by taking advantage of the easy sharding and replication provided by ReDB.

**Comment by lucasjans** [comment]
We are experiencing similar behavior with much smaller documents in #5695
